Neos.Flow/Classes/ResourceManagement/ResourceTypeConverter.php

```diff
@@ -292,26 +311,28 @@ protected function handleUploadedFile(UploadedFileInterface $source, PropertyMap
             case \UPLOAD_ERR_INI_SIZE:
             case \UPLOAD_ERR_FORM_SIZE:
             case \UPLOAD_ERR_PARTIAL:
-                return new FlowError\Error(Files::getUploadErrorMessage($source->getError()), 1264440823);
+                return new FlowError(Files::getUploadErrorMessage($source->getError()), 1264440823);
             default:
                 $this->logger->error(sprintf('A server error occurred while converting an uploaded resource: "%s"', Files::getUploadErrorMessage($source['error'])), LogEnvironment::fromMethodName(__METHOD__));
 
-                return new FlowError\Error('An error occurred while uploading. Please try again or contact the administrator if the problem remains', 1340193849);
+                return new FlowError('An error occurred while uploading. Please try again or contact the administrator if the problem remains', 1340193849);
         }
 
         if (isset($this->convertedResources[spl_object_hash($source)])) {
             return $this->convertedResources[spl_object_hash($source)];
         }
 
         try {
-            $resource = $this->resourceManager->importResource($source->getStream()->detach(), $this->getCollectionName($source, $configuration));
+            $resource = new PersistentResource($source->getStream()->detach());
```

Comment:
That's a (kind of big) behavior change isn't it? If I use this converter I could so far expect it to create a persisted resource.

Reply:
Well, it is an `"eventually persisted resource"` so to say. I'd be happy to find another solution to allow validation to prevent persisting (and importing) a resource, but currently the property mapping is done *before* validation. And this is not a feature, but actually a bug. A Resource that doesn't pass validation should not end up in the database, let alone on persistent storage. But it's also not easy to just remove the resource from persistence again after validation, as you'd need to know if the resource was newly added or just updated (plus the unnecessary work done). So the converter really should not call `repository->add()` (the `PersistentObjectConverter` doesn't either), nor move the resource to a storage backend.

The only valid other option I see would be to split validation in two separate steps, then only have the first step allow rejecting resources (in their "source" form). But that would be one hell of a change in regards to code required, complexity, breakyness and required documentation to make people understand what the two validations are.

Anyway: Thanks for challenging the solution!

Reply:
PS: This whole converter is the reason I did not introduce a `EventuallyPersistedResource extends PersistentResource` class - because the converter would need to change, which would be super-breaking. With my solution this is more subtle and a `PersistentResource` is differentiated by the `source` property. One with a source has definitely not yet been imported! It of course won't work if you manually create a `new PersistentResource($someSource)` and add that to a repo. But that also didn't work before, you needed to use the `ResourceManager` and if you do that, this solution right now also works.
We could maybe also add a safety-check into the `ResourceRepository` that allows only adding resources that have no source attached, but then again, that would only make it better for those cases where you specify a `$source` - which in turn would be fine in scope of this change.